### PR TITLE
Add a `delayed_pure` global setting

### PR DESCRIPTION
**Pull request by jcrist**
Adds `delayed_pure` to `set_options`, and changes the precedence of
`pure` in delayed calls to the following (in order of decreasing
precedence):

- Specified via the keyword `pure` in the call directly (e.g. `foo(1, 2,
  pure=True`))
- Specified as a property of the callable via the `pure` keyword to
  `delayed` (e.g. `delayed(foo, pure=True)(1, 2)`)
- Specified via the global `delayed_pure` setting
- Otherwise False

Supersedes #2169.

### CODE REVIEW

**Comment by jcrist** [comment]
This test should help demonstrate behavior:

```python
def test_pure_global_setting():
    # delayed functions
    func = delayed(add)

    with set_options(delayed_pure=True):
        assert func(1, 2).key == func(1, 2).key

    with set_options(delayed_pure=False):
        assert func(1, 2).key != func(1, 2).key

    func = delayed(add, pure=True)
    with set_options(delayed_pure=False):
        assert func(1, 2).key == func(1, 2).key

    # delayed objects
    assert delayed(1).key != delayed(1).key
    with set_options(delayed_pure=True):
        assert delayed(1).key == delayed(1).key

    with set_options(delayed_pure=False):
        assert delayed(1, pure=True).key == delayed(1, pure=True).key

    # delayed methods
    data = delayed([1, 2, 3])
    assert data.index(1).key != data.index(1).key

    with set_options(delayed_pure=True):
        assert data.index(1).key == data.index(1).key
        assert data.index(1, pure=False).key != data.index(1, pure=False).key

    with set_options(delayed_pure=False):
        assert data.index(1, pure=True).key == data.index(1, pure=True).key
```

cc @ordirules, @mrocklin for thoughts.

File: dask/delayed.py [hunk 1]
```diff
@@ -437,7 +461,7 @@ def call_function(func, func_token, args, kwargs, pure=False, nout=None):
 class DelayedLeaf(Delayed):
     __slots__ = ('_obj', '_key', '_pure', '_nout')
 
-    def __init__(self, obj, key, pure=False, nout=None):
+    def __init__(self, obj, key, pure=None, nout=None):
         self._obj = obj
         self._key = key
         self._pure = pure
```
Comment (ordirules) [inline dask/delayed.py]: This is the same comment as a previous line note. In DelayedAttr below, it seems purity is always assumed True. I'm fine with it if you are okay with it (and @mrocklin ), but maybe it should be documented? (There could be cases like an attribute spitting out a random number. We have image series simulators where indexing an image in the image sequence generates a new random image for example. Of course we could implement this more explicitly for a task graph, so long as we are properly warned about it.)

**Comment by shoyer** [comment]
I haven't looked at this carefully -- but do you really need the global state provided by options here? Would it work to simply provide provide alternative constructors/classes?

Something like `from dask import always_pure_delayed as delayed` is far more explicit than using `set_options()`.

**Comment by jcrist** [comment]
> I haven't looked at this carefully -- but do you really need the global state provided by options here? Would it work to simply provide provide alternative constructors/classes?

The point of this is to make operations pure/impure inside a context. Specifying purity on method calls can be currently done by adding `pure=True` to the method call, but it'd be nice to have a way to avoid having to do this. `delayed` is primarily intended for wrapping functions, but of course can work with any callable. How do you specify that the methods on the output of a callable are also pure? Or the methods on the results of those methods? Having a reasonable set of rules for this is tricky, it seems easier/clearer (to me) to just do so on a context basis.

```python
@delayed(pure=True)
def foo(x):
    return [1] * x

list_of_ones = foo(10)
# Methods are impure by default
assert list_of_ones.count(1).key != list_of_ones.count(1).key
# Can specify purity in the call
assert list_of_ones.count(1, pure=True).key == list_of_ones.count(1, pure=True).key
# Can specify purity in context
with dask.set_options(delayed_pure=True):
    assert list_of_ones.count(1).key == list_of_ones.count(1).key
```

Note that this is *not* about avoiding the need to specify `pure` in the decorator/wrapper. Since `delayed` is a curried function, this is already easy:

```python
always_pure_delayed = delayed(pure=True)
```

File: dask/delayed.py [hunk 2]
```diff
@@ -267,11 +288,14 @@ def delayed(obj, name=None, pure=False, nout=None, traverse=True):
...
+    As with function calls, method calls also respect the global
+    ``delayed_pure`` setting and support the ``dask_key_name`` keyword:
 
     >>> a.count(2, dask_key_name="count_2")
-    Delayed("count_2")
+    Delayed('count_2')
+    >>> with dask.set_options(delayed_pure=True):
+    ...     print(a.count(2).key == a.count(2).key)
+    True
     """
```
Comment (jcrist) [inline dask/delayed.py]: This is true, but it's already documented. Is this not clear enough? Suggestions for improvement? https://github.com/dask/dask/blob/master/dask/delayed.py#L243-L246.

Comment (ordirules) [inline dask/delayed.py]: no improvement. i apologize i missed this!
looks good!

**Comment by shoyer** [comment]
> How do you specify that the methods on the output of a callable are also pure? Or the methods on the results of those methods? 

If all you need is a way to say that all method calls and results will be pure until explicitly signaled otherwise, then it seems like it would make sense to make purity an enum rather than a boolean, with values for "impure", "pure once" and "pure always". It's clearly not obvious what `pure=True` means for methods methods, so using an enum is a good way to force people to think about this.

More generally, I'm not a big fan of context managers for controlling options, because they can have hard to predict non-local consequences. If there are other good options, they should be preferred.

Anyways... I'll let you think about this and make a decision. I don't really have a horse in this fight.

**Comment by mrocklin** [comment]
I view context managers or globals as a way to make our default choices more programmable.  

For code maintainability reasons we don't pass around all of the options all the time.  Instead we have default behaviors.  As people want to change a few of these default behaviors we are asked to significantly expand the API for the support of corner cases.  I agree with @shoyer that context managers generally has costs over specifying a different kind of function or different PureDelayed class or whatnot, but we would probably refuse to include and maintain all of the various options here as separate functions/classes whatever.  Using context managers is a hack, but a hack that is easy to build and maintain.  I view it as a release valve for users with more corner-case needs until we find the one-true-solution.

**Comment by mrocklin** [comment]
It's also worth noting that people can subclass Delayed on their own.

File: dask/tests/test_delayed.py [hunk 3]
```diff
@@ -213,6 +214,40 @@ def test_pure():
...
+    # delayed objects
+    assert delayed(1).key != delayed(1).key
+    with set_options(delayed_pure=True):
+        assert delayed(1).key == delayed(1).key
+
+    with set_options(delayed_pure=False):
+        assert delayed(1, pure=True).key == delayed(1, pure=True).key
+
+    # delayed methods
+    data = delayed([1, 2, 3])
```
Comment (ordirules) [inline dask/tests/test_delayed.py]: `assert data.index.key == data.index.key` maybe?
(magic method attribute always pure test)

File: dask/tests/test_delayed.py [hunk 4]
```diff
@@ -213,6 +214,40 @@ def test_pure():
...
+
+    with set_options(delayed_pure=False):
+        assert func(1, 2).key != func(1, 2).key
+
+    func = delayed(add, pure=True)
+    with set_options(delayed_pure=False):
+        assert func(1, 2).key == func(1, 2).key
+
+    # delayed objects
+    assert delayed(1).key != delayed(1).key
```
Comment (ordirules) [inline dask/tests/test_delayed.py]: `assert (delayed(1)+delayed(1)).key == (delayed(1) + delayed(1)).key` 
(magic method binary op always pure test)

**Comment by ordirules** [comment]
Actually, one question, how will this fare in a distributed environment? Anything accessing the scheduler will have to agree upon a global state?

**Comment by mrocklin** [comment]
What is the status of this?

**Comment by jcrist** [comment]
> What is the status of this?

This should be good to merge. Seems to be an unrelated failure introduced in https://github.com/dask/dask/pull/2163.

> Actually, one question, how will this fare in a distributed environment? Anything accessing the scheduler (and hence workers) will have to agree upon a global state?

The setting here only affects graph building (which is what `delayed` is for). During compute, the schedulers only see the graph, and don't know anything about delayed or any of these settings. So yes, this works fine in a distributed setting.

**Comment by ordirules** [comment]
@jcrist thanks for the information. I get the idea better now. I've been reading more and more about it and I think I grasp it better (sorry for the previously naif questions I could have answered by more thorough reading. :-( ). This will definitely help us.

One aside question. we want stuff such as `foo(1,b=2)` and `foo(1,2)` to hash to the same key. This seems to be as simple as an extra decorator + the `inspect` module to figure out if `2` is a pure positional argument or a named argument and to change, for example `foo(1,2)` to `foo(1,b=2)` before it's delayed (forgive my nomenclature, I'm not quite sure what to call each... using `kwarg` seems to meant something else).
I'm just curious if dask intend to eventually support something like this? (Or if perhaps something like the very neat `toolz` library has something like this I may have missed).

thanks!

**Comment by jcrist** [comment]
> One aside question. we want stuff such as `foo(1,b=2)` and `foo(1,2)` to hash to the same key.

I'm going to say this is outside the scope of `dask.delayed`, as there are too many edge cases that would make doing this generally in a way that is also performant tricky.

I'm not sure what your use is, but you could do something like this to ensure things are always arguments/keyword arguments:

```python
def user_facing(a, b=2):
    return actual_task(a, b)

@delayed(pure=True)
def actual_task(a, b):
    ...
```

**Comment by jcrist** [comment]
Unrelated failure was fixed, all tests passing. Merging.

**Comment by ordirules** [comment]
yep, that's what i meant by decorator +inspect (so it can be done generally)

thanks! :-)

**Comment by ordirules** [comment]
@jcrist is it possible to do something similar with `tokenize`?
I have objects/dictionaries that contain attributes not necessary to the computation that I don't want to hash. In that case it would be convenient to define my own tokenize routine and use it.

Basically, each result from a dask computation is a dictionary that I also assign a uid. I need it to be unique (i.e. if i stop and restart python, it needs to be a unique number). But I don't want it to hash with the data since it will always randomize the hash (which is the same as `pure=False` practically).

I imagine there could be other cases out there where users may want to define their own tokenize routines as well. Any thoughts/ideas would be appreciated. (I could implement this on my own as well independent of dask)

I can post as issue for better tracking if you think it's worth it. thanks!

**Comment by ordirules** [comment]
Thanks, I think `dask_key_name` will suffice if needed. I was wondering if there was a way (or plans) to allow user to override the usual tokenization.

However, that said... There was a flaw in my idea. I'll try to explain anyway  below this message, in case you are interested (or have some feedback). Thanks for your input, this is much appreciated.

What I had meant can maybe be illustrated by the following (very crude) example:
```python
@delayed(pure=True)
def sumcounts(result, counts):
    # add counts to counts element in result, and return a new result with new uid
    newresult = result.copy()
    newresult['counts'] += counts
    # this piece is impure, but for purposes of the calculation can be ignored (and assumed pure)
    newresult['uid'] = uuid4()
    return newresult

myresult = dict()
myresult['uid'] = uuid4()
myresult['counts'] = 0

# myresult contains a random uid, but I don't want that to be tokenized since the computation
# in itself is pure
myresult2 = sumcounts(myresult, 3)

# some time later....
anotherresult = dict()
anotherresult['uid'] = uuid4()
anotherresult['counts'] = 0

# this should not be re-computed, even though uid's are different
anotherresult2 = sumcounts(anotherresult, 3)
```
The problem with this is that if `anotherresult2` is not computed and retrieved from cache, it won't contain a unique uid. I'm not sure how to actually implement this without either dropping the uid-like terms or adding attributes on the cache-level (which dask seems very friendly for user created caches). I'm dropping it for now, but will be looking into the cache-alternative (maybe use something like zict too, which would allow writing to disk).
